### ISSUE TITLE
Register a dbt package from a sub directory of a repository 

### ISSUE THREAD

**Issue by turbo1912** [dbt-labs/hubcap#95]
Hi friends, 

We are about to release a new dbt package and we wanted to keep everything as a morepo in our project. That would mean instead of pointing to a github repository we would have to point it to a subdirectory in our existing repository for example; https://github.com/fal-ai/fal/feature-store. 

I don't think this is possible right now, but we are happy to contribute if this is something you would like to see implemented. 

I just had a quick glance at the code and looks like a change like this is self contained in the hubcap repository and I don't have to touch the logic how `dbt-core` downloads dependencies.

**Comment by elongl** [comment]
We'd really love this feature as well!
Will also like to contribute if possible.

**Comment by joellabes** [comment]
@turbo1912 and @elongl thanks for opening this, and sorry that it took a while to get back to you! 

Conceptually, I'm very supportive of this, especially if it doesn't impact how dbt-core downloads things. 

The things that are giving me pause are that we don't have any testing around the project (other than a couple of semver tests that @dbeatty10 added in https://github.com/dbt-labs/hubcap/pull/133) at the moment so can't really guarantee that changes don't cause downstream issues. 

I would say go for it, with the disclaimer that code review might be slow as we try to rustle up someone who knows how hubcap works. We'll also probably wind up dragging in someone from @dbt-labs/core to double check that there aren't any flow-on effects.

**Comment by elongl** [comment]
Awesome! Glad to hear you and the team are on board.
I'll send here in a couple of days a rough design of how I'm planning to implement it so I could get your approval and start working on it. Thanks a lot.

**Comment by jtcohen6** [comment]
I definitely see the value here!

I think this may be tricky for us to do in the current implementation of Hubcap + Hub site (hub.getdbt.com), though not impossible. The Hub site does not actually store/mirror specific files—it's just a pointer to a GitHub tarball URL, containing a zipped version of _all_ files from the repo. E.g. for `dbt_utils` version 0.8.6: https://codeload.github.com/dbt-labs/dbt-utils/tar.gz/0.8.6

If we wanted to proceed with this as an extension of the current implementation, I think it would need to include:
- A new Hub API field, `subdirectory`
- An update to some methods in `dbt-core`'s `deps` logic ([`download_and_untar`](https://github.com/dbt-labs/dbt-core/blob/b43fc76701769d41340184d4dfab04bcb38646f3/core/dbt/deps/registry.py#L76-L86), [`untar_package`](https://github.com/dbt-labs/dbt-core/blob/b43fc76701769d41340184d4dfab04bcb38646f3/core/dbt/clients/system.py#L474-L483)) to accept a `subdirectory` parameter, pull + rename only that subdirectory if a `subdirectory` argument is supplied, and delete the remaining files (! will require careful testing on any OS with odd file permissions, a.k.a. Windows)

Note that will still require downloading the entire contents of the repo, and then quickly deleting the contents we don't care about. That could still pose a risk on containerized / disk-limited file systems, if the overall repo is truly massive.

A longer-term answer probably looks like the Hub graduating to support its own file-hosting capabilities, rather than using GitHub as a backend. In that future, something like this should be much simpler to implement, and better: the filtering can happen during package registration / upload, rather than in every single package download.

Related issue: https://github.com/dbt-labs/dbt-core/issues/4868

**Comment by elongl** [comment]
Hi @jtcohen6 
Really happy to see that you support the idea.
I think I'll begin working on it this weekend and submit a PR.

Too bad Github doesn't provide a way to only download a specific file path within a repository 😞 
Where should the user specify the `subdirectory` within the API, especially the `hub.json`?

I imagined it as something like that:
```json
    "elementary-data": [
        "dbt-data-reliability/<subdirectory>"
    ]
```
so basically if there's a subdirectory specify the full path within the repository with leading `/`s.
What do you think?

**Comment by lostmygithubaccount** [comment]
would using `git-sparse-checkout` help? https://git-scm.com/docs/git-sparse-checkout. It does allow you to effectively only download a subdirectory, I'm not sure how easy it is to use though

**Comment by jtcohen6** [comment]
We use "sparse checkout" for package subdirectories installed via the `git` method: https://docs.getdbt.com/docs/building-a-dbt-project/package-management#project-subdirectories

The installation mechanism for the `package` method (= Hub registry packages) doesn't actually use `git`, though — it just downloads tarball contents from the URL supplied by GitHub's API (`codeload.github.com`).

**Comment by joellabes** [comment]
A pleasant side effect: if this _was_ done, [experimental](https://github.com/dbt-labs/dbt-utils/discussions/487#:~:text=I%27ll%20hear%20it!-,Experiments,-This%20is%20not) packages (such as insert_by_period) could stay on the hub instead of needing to be specified as git subdirectories

**Comment by elongl** [comment]
Bumping [this](https://github.com/dbt-labs/hubcap/issues/95#issuecomment-1200390533) before I'm starting to work on it.
Would appreciate your confirmation, thanks!

**Comment by elongl** [comment]
@jtcohen6 @joellabes 
Bumping [this comment](https://github.com/dbt-labs/hubcap/issues/95#issuecomment-1200390533) before I'm starting to work on it.
Would appreciate your confirmation, thanks!

**Comment by elongl** [comment]
@jtcohen6 Yes, of course! It also makes much more sense to me with a slight change.
I think it's better design to keep the organization within the key rather than specify it everytime.
```json
{
    "organizations": {
        "elementary-data": {
            "packages": [
                {
                    "repo_name": "dbt-data-reliability",
                    "subdirectory": "dbt_project"
                }
            ]
        }
}

**Comment by jtcohen6** [comment]
@elongl Fair point! I can't think of another org-level property that we'd need to specify.

We could opt for as much conciseness as possible, while still allowing for structure where we need it. The org name could be the key, and its value would be a list of packages, with type `List[str, Dict[str, str]]`:
```json
{
    "elementary-data": [
        "dbt-repo-non-subdirectory",
        {
          "repo_name": "dbt-data-reliability",
          "subdirectory": "dbt_project"
        },
    ]
}

```

---

I'm realizing this would also offer a (roundabout) way of resolving https://github.com/dbt-labs/dbt-core/issues/4868 (a way to ignore/exclude large unnecessary files). Package maintainers could move the "essential" components of the package to a subdirectory, and then specify that subdirectory in `hub.json`. In effect, `dbt-core` would ignore (not copy) / delete the other files. It's not a perfect solution, but it feels like a reasonable workaround, until we have our own dedicated infrastructure for hosting package files.

**Comment by domenic-donato** [comment]
We're also looking to setup this tool in our monorepo. Is this now possible? If so, what are the steps we should follow.

**Comment by dbeatty10** [comment]
Thanks for letting us know your interest @domenic-donato.

This is still a feature request that hasn't been implemented or released.
